File: engine/camera.py
class Camera:
    def __init__(self):
        self.x = 0.0
        self.y = 0.0
        self.zoom = 1.0

        self.min_zoom = 0.5
        self.max_zoom = 2.0

        self.bounds = None
        self.viewport = {
            "x": 0,
            "y": 0,
            "width": 1,
            "height": 1,
        }

    def set_bounds(self, x, y, width, height):
        self.bounds = {
            "x": x,
            "y": y,
            "width": width,
            "height": height,
        }
        self.clamp()

    def set_viewport(self, rect):
        self.viewport = {
            "x": int(rect.x),
            "y": int(rect.y),
            "width": max(1, int(rect.width)),
            "height": max(1, int(rect.height)),
        }
        self.clamp()

    def move(self, dx, dy):
        self.x += dx
        self.y += dy
        self.clamp()

    def set_zoom(self, zoom):
        self.zoom = max(self.min_zoom, min(self.max_zoom, zoom))
        self.clamp()

    def zoom_by(self, amount):
        self.set_zoom(self.zoom + amount)

    def world_to_screen(self, wx, wy):
        sx = (wx - self.x) * self.zoom + self.viewport["x"]
        sy = (wy - self.y) * self.zoom + self.viewport["y"]
        return sx, sy

    def screen_to_world(self, sx, sy):
        wx = (sx - self.viewport["x"]) / self.zoom + self.x
        wy = (sy - self.viewport["y"]) / self.zoom + self.y
        return wx, wy

    def clamp(self):
        if not self.bounds:
            return

        self.x = max(self.bounds["x"], self.x)
        self.y = max(self.bounds["y"], self.y)

        view_width = self.viewport["width"] / max(self.zoom, 0.001)
        view_height = self.viewport["height"] / max(self.zoom, 0.001)

        max_x = max(self.bounds["x"], self.bounds["width"] - view_width)
        max_y = max(self.bounds["y"], self.bounds["height"] - view_height)

        self.x = min(self.x, max_x)
        self.y = min(self.y, max_y)

File: src/engine/camera.rs
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Camera {
    pub x: f64,
    pub y: f64,
    pub zoom: f64,
    pub min_x: f64,
    pub min_y: f64,
    pub max_x: f64,
    pub max_y: f64,
    pub viewport: (f64, f64, f64, f64),
}

impl Default for Camera {
    fn default() -> Self {
        Self {
            x: 0.0,
            y: 0.0,
            zoom: 1.0,
            min_x: 0.0,
            min_y: 0.0,
            max_x: 0.0,
            max_y: 0.0,
            viewport: (0.0, 0.0, 1100.0, 740.0),
        }
    }
}

impl Camera {
    pub fn set_bounds(&mut self, min_x: f64, min_y: f64, max_x: f64, max_y: f64) {
        self.min_x = min_x;
        self.min_y = min_y;
        self.max_x = max_x;
        self.max_y = max_y;
        self.clamp_to_bounds();
    }

    pub fn set_viewport(&mut self, viewport: (f64, f64, f64, f64)) {
        self.viewport = viewport;
    }

    pub fn move_by(&mut self, dx: f64, dy: f64) {
        self.x += dx;
        self.y += dy;
        self.clamp_to_bounds();
    }

    pub fn set_zoom(&mut self, zoom: f64) {
        self.zoom = zoom.clamp(0.1, 8.0);
    }

    pub fn clamp_to_bounds(&mut self) {
        if self.max_x > self.min_x {
            self.x = self.x.clamp(self.min_x, self.max_x);
        }
        if self.max_y > self.min_y {
            self.y = self.y.clamp(self.min_y, self.max_y);
        }
    }
}
